aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -31,30 +31,47 @@
 namespace at {
 namespace native {
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eye ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
 Tensor& eye_out_cuda(int64_t n, Tensor& result) {
   // the default value of `m` equals to `n`
   return at::native::eye_out_cuda(n, n, result);
 }
 
 Tensor& eye_out_cuda(int64_t n, int64_t m, Tensor& result) {
+  // the default value of `k` equals to 0
+  return at::native::eye_out_cuda(n, m, 0, result);
+}
+
+Tensor& eye_out_cuda_nk(int64_t n, c10::optional<int64_t> k, Tensor& result) {
+  // the default value of `m` equals to `n`
+  int64_t k_v{k.has_value() ? k.value() : 0};
+  return at::native::eye_out_cuda(n, n, k_v, result);
+}
+
+Tensor& eye_out_cuda(int64_t n, int64_t m, int64_t k, Tensor& result) {
   TORCH_CHECK(n >= 0, "n must be greater or equal to 0, got ", n);
   TORCH_CHECK(m >= 0, "m must be greater or equal to 0, got ", m);
+  TORCH_CHECK((k > -n && k < m) || (k == 0),
+              "k out of range, should be between (-",
+              n, ", ", m, "), or 0 but got ", k);
 
   result.resize_({n, m});
   result.zero_();
+  result.diagonal(k).fill_(1.);
```

Comment:
It's probably a bug that this works for integer tensors (an inplace add of an integer tensor with a float scalar will, for example, throw an error). Let's just change `1.` to `1` to be more future proof?